lib/stream.ml

```diff
@@ -277,11 +285,59 @@ let emitter ?(len = 65535 * 4) () =
   let len = Bytes.length buf |> Unsigned.Size_t.of_int in
   B.emitter_set_output_string e (Ctypes.ocaml_bytes_start buf) len written;
   match r with
-  | 1 -> Ok { e; event; written; buf }
+  | 1 -> Ok { e; event; kind = String { written; buf } }
+  | n -> Error (`Msg ("error initialising emitter: " ^ string_of_int n))
+
+let emitter_file file =
+  let e = Ctypes.(allocate_n T.Emitter.t ~count:1) in
+  let event = Ctypes.(allocate_n T.Event.t ~count:1) in
+  let r = B.emitter_init e in
+  B.emitter_set_output_file e file;
+  match r with
+  | 1 -> Ok { e; event; kind = File }
+  | n -> Error (`Msg ("error initialising emitter: " ^ string_of_int n))
+
+let emitter_handler handler =
+  let e = Ctypes.(allocate_n T.Emitter.t ~count:1) in
+  let event = Ctypes.(allocate_n T.Event.t ~count:1) in
+  let r = B.emitter_init e in
+  let open Ctypes in
+  let handler _ buf len =
+    let buf' = Ctypes.(coerce (ptr uchar) (ptr char) buf) in
+    let s =
+      Ctypes.(string_from_ptr buf' ~length:(Unsigned.Size_t.to_int len))
+    in
+    handler s;
+    1
+  in
+  let f =
+    coerce
```

Comment:
I don't understand how the coercion could be correct.

Reply:
Maybe @yallop could confirm whether this is a correct way to pass ocaml callback on the C side (note that the callback will be called after the C function call returns).

Reply:
I think I just based it on https://github.com/yallop/ocaml-ctypes/issues/364#issuecomment-258612489.